bst.py

```diff
@@ -165,7 +166,131 @@ def breadth_first(self):
             if len(list_of_nodes_at_next_depth) == 0:
                 break
 
+    def left(self, node):
+        return self.nodes[node].get('left')
+
+    def right(self, node):
+        return self.nodes[node].get('right')
+
+
+    def delete(self, val):
+        '''Remove val from the tree if present, return None in all cases.'''
+        print 'value: {}\nleft: {}\nright: {}'.format(val, self.nodes[val].get('left'), self.nodes[val].get('right'))
+        print 'self.start: {}'.format(self.start)
+        if val not in self.nodes:
+            return
+        # No children
+        elif self.nodes[val].get('left') is None and self.nodes[val].get('right') is None:
+            if self.size() == 1:
+                self.start = None
+                del self.nodes[val]
+            else:
+                self._change_ref(val)
+                del self.nodes[val]
+            return
+        # Single child
+        left = self.nodes[val].get('left')
+        right = self.nodes[val].get('right')
+        # The case where only a right child exists.
+        if left is None:
+            self._change_ref(val, right)
+            del self.nodes[val]
+        elif right is None:
+            self._change_ref(val, left)
+            del self.nodes[val]
+        # 3rd
+        elif val > self.start:  # on right side of tree
+            temp = self.nodes[left]
+            self._change_ref(val, left)
+
+            self.nodes[left]['right'] = self.nodes[val]['right']
+            self.nodes[right]['parent'] = left
+
+            self.nodes[val]['left'] = temp.get('left')
+            self.nodes[val]['right'] = temp.get('right')
+            self.nodes[val]['parent'] = left
+            self.delete(val)
```

Comment:
I know this is the consequence of the dict storage thing, but man this code is hard to look at and parse.  I imagine you feel the same.  I'd love to see some of these sets of lines collapsed to helper functions so that they were more clearly parseable.